Native by default

DemiMarie opened the pull request:
Currently the "obvious" choice for the OCaml tools are ocamllex,
ocamlc, and ocamlopt.  However, there is no good reason to use them
when ocamllex.opt, ocamlc.opt, and ocamlopt.opt are available.

Therefore, install the native-code versions with no suffixes when
they are present.  The bytecode versions are now available with
the `.byte` suffix.

DemiMarie committed:
No suffix for ocamlc.opt/ocamlopt.opt/ocamllex.opt

Currently the `obvious` choice for the OCaml tools are ocamllex,
ocamlc, and ocamlopt.  However, there is no good reason to use them
when ocamllex.opt, ocamlc.opt, and ocamlopt.opt are available.

Therefore, install the native-code versions with no suffixes when
they are present.  The bytecode versions are now available with
the `.byte` suffix.

DemiMarie committed:
Add Changes entry

dbuenzli commented:
xref [MPR 7014](http://caml.inria.fr/mantis/view.php?id=7014).

gasche commented:
Just to clarify: the proposed change does what one would expect, that is install the bytecode tool as `foo.byte`, but install the native compiler in _both_ `foo` and `foo.opt` (for backward-compatibility).

Naive question: Is it safe to use `ln -s` to install files? I wonder how Windows handles that for example. Looking at the existing makefiles, the uses where either
- internals to the compiler build tree, or
- for [the target installopt-noprof](7fd3559990683f4f0acdb89d4bedcbe4fe7ff631) that I don't anyone ever uses in practice, or 
- coming from a [recent change](0225ca01e39289ce1801fb09cd011cdbfb542b8d)

so we cannot assume that this has been tested before.

DemiMarie committed:
Added my name to `Changes`

DemiMarie committed:
Don't use `ln -sf` on Windows

DemiMarie commented:
Yes, the bytecode tool is now `foo.byte`.  The native tool is now in both `foo` and `foo.opt`, if it has been built.  If the native tool has not been built, the bytecode tool is installed as `foo`.

The idea is that running `foo` will automatically run a native-code tool if available, or a bytecode tool otherwise.  This makes it easy to write scripts that invoke the optimum OCaml tool cross-platform.

I got rid of the use of `ln -s` on Windows

alainfrisch commented:
If there is no reason to use the bytecode version, why do we install them at all?

mshinwell commented:
I have at least two use cases for installing bytecode.
1. (By example.) I'm working on a version of the compiler that adds certain instrumentation for native code.  This has to be enabled at compiler configuration time, since the compiler and build system do not currently support a scheme where selecting a compiler flag at runtime causes objects to be generated that are incompatible with objects generated without that flag.  It would be lovely if this would work, but it requires real effort to make this work properly.  All of the libraries (including the ones in otherlibs/, not just the stdlib) need to be built multiple times and there has to be a reasonable way of ensuring the correct ones are linked against when the user just writes e.g. "threads.cmxa" on the command line.  Without this support, what actually happens is the compiler itself ends up being built with instrumentation, which in this case cannot then be easily disabled at runtime.  This may pose problems both for runtime performance and memory consumption.  In such a scenario it might be useful to be able to run the ocamlopt that executes on the bytecode interpreter, since it avoids the instrumentation.
2. Production of ocamlopt.opt involves significantly more effort on the part of the compiler than production of ocamlopt (especially in Flambda mode), and thus there is a higher chance of a miscompilation due to a bug.  In the fairly unlikely but possible scenario this happens, it may be possible to use the bytecode ocamlopt instead as a temporary workaround.

I don't really see the small overhead incurred by installing the bytecode versions to be a problem, and would suggest erring on the side of caution in any case.

dbuenzli commented:
> Production of ocamlopt.opt involves significantly more effort on the part of the compiler than production of ocamlopt (especially in Flambda mode), and thus there is a higher chance of a miscompilation due to a bug. In the fairly unlikely but possible scenario this happens, it may be possible to use the bytecode ocamlopt instead as a temporary workaround.

This suggest that we rather want to be able to `configure` which compilers get installed. From a build system perspective I'd like to be able not to be concerned about `.opt` or `.byte` suffixes and simply invoke a tool without extensions. Whether this is a native or byte code version should not be of concern to the build system and the build system should be able to assume that the tool it invokes will be the best choice under the situation at hand.

DemiMarie committed:
Install native-code versions of remaining tools

DemiMarie committed:
Fix build bustage and delete duplicate code

DemiMarie committed:
Fix syntax error in shell code

DemiMarie commented:
I have changed the makefile so that all of the installed tools that are built in the `tools/` directory are now installed in both bytecode and (where available) native code.

Should I squash the commits?  The history is 16 commits long and some of them involve trivial typos.

DemiMarie committed:
Fix bug resulting in bogus symlinks being installed

DemiMarie committed:
Filter out unwanted files from installation

DemiMarie committed:
Fix Windows install

DemiMarie committed:
Fix Windows install

DemiMarie commented:
@damiendoligez any thoughts?

gasche commented on tools/Makefile.shared:
```diff
@@ -12,10 +12,53 @@
 include ../config/Makefile
-CAMLRUN ?= ../boot/ocamlrun
-CAMLYACC ?= ../boot/ocamlyacc
+ifndef CAMLRUN
```
`$$$$` looks a bit scary. Could you maybe add a comment on what the subtle code below is doing and why it is necessary?

DemiMarie replied:
Done.

DemiMarie committed:
Make install paths with whitespace slightly less broken on *nix

DemiMarie committed:
Document the subtle code in tools/Makefile.shared

DemiMarie committed:
More documentation of tools/Makefile.shared

gasche commented on tools/primreq.ml:
```diff
@@ -40,16 +40,15 @@ let scan_info cu =
 let scan_obj filename =
   let ic = open_in_bin filename in
-  let buffer = String.create (String.length cmo_magic_number) in
+  let buffer = Bytes.create (String.length cmo_magic_number) in
```
You should replace this and the following `really_input` by a single call to `really_input_string`, which allows to avoid the unsafe coercions below. If you do that and rebase the changes to this file in a single commit at the bottom of your commit stack (close to master), I'll push it to master now.

DemiMarie replied:
Can you explain how to do this?  This is really hard in git.

SGrondin replied:
@drbo There are some good tutorials online about `git rebase --interactive`. The high level explanation is that since your branch and the branch you want to merge into have both evolved independently, there'll be conflicts and the merge is guaranteed to be messy. By doing a rebase, git reapplies each of the commits in your branch on top of the branch to merge into, so you get to deal with conflicts one at a time and even do some cleanup like merging some of your commits together. Then once that's done you can open a Pull Request and it's easier to review and the merge is clean and painless. I can help you with this if you have Skype or Hangouts.

gasche replied:
So here `git rebase --interactive master` is the command to start from (after you have pushed a last patch implementing the change in this file that you want to merge with the current diff). Feel free to accept @SGrondin's offer for additional explanations, otherwise just add a patch and I can rebase it on my side.

DemiMarie committed:
Update .depend and add -compat-32 to ocamlc calls

DemiMarie committed:
Change references to `objinfo` to `ocamlobjinfo`

Address review comments and clean up quoting

Fix inconsistent makefile quoting and testsuite bustage

Fix includes of makefiles

Fix configure bug

make travis rerun

Trying to debug travis issue

Try to fix configure on Travis

Add config.mk which I had previously forgotten

to commit.

Fix silly makefile bug

Fix bug causing STUBLIBDIR to be empty

Fix makefile quoting bugs

Fix MSVC build break

DemiMarie commented:
Testsuite failure is same as trunk.

gasche commented:
I pushed the `really_input_string` change in tools/primreq.

damiendoligez commented on Makefile:
```diff
@@ -198,11 +198,13 @@ base.opt:
-INSTALL_MANDIR=$(DESTDIR)$(MANDIR)
+override shellquote = ""'''$(subst ','\'',$1)'#)'
+
+INSTALL_BINDIR:=$(call shellquote,$(DESTDIR)$(BINDIR))
```
Why do we need to shellquote these variables and not directly DESTDIR, BINDIR, etc?
Why should it be done in the Makefile rather than the configure script?

damiendoligez commented on Makefile:
```diff
@@ -215,76 +217,85 @@ install:
 	  else $(MKDIR) $(INSTALL_MANDIR)/man$(MANEXT); fi
-	cp VERSION $(INSTALL_LIBDIR)/
-	cd $(INSTALL_LIBDIR); rm -f \
+	cp -- VERSION $(INSTALL_LIBDIR)/
```
Interestingly enough, the `--` option is not documented for `cp` or `cd`. Have you checked that it's portable across all our supported architectures? (Think of the BSDs!)

Moreover, this change will be superseded by [MPR#5219](http://caml.inria.fr/mantis/view.php?id=5219) at some point.

damiendoligez commented on byterun/Makefile.common:
```diff
@@ -93,8 +93,8 @@ install-runtimed:
 install::
-	cp ocamlruni$(EXE) $(INSTALL_BINDIR)/ocamlruni$(EXE)
-	cp libcamlruni.$(A) $(INSTALL_LIBDIR)/libcamlruni.$(A)
+	cp ocamlruni$(EXE) "$(INSTALL_BINDIR)/ocamlruni$(EXE)"
```
You're using a different mechanism for quoting `INSTALL_BINDIR` here and in `Makefile`. That can't be right.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -1,4 +1,4 @@
-#**************************************************************************
+#************************-*-makefile-*-************************************
```
You can't change the header format without talking to me first. See `tools/check-typo`.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -12,10 +12,62 @@
 #*                                                                        *
 #**************************************************************************
-
+MAKEFLAGS := -r -R
```
What's the point? You're not doing any recursive `make` invocation in this directory.

DemiMarie replied:
GNU make 4.0+ honors `MAKEFLAGS` even for the current invocation of `make`.

damiendoligez replied:
Then I have another question: why do you need to remove the default rules and variables, and what does it have to do with installing the native version of the tools?

BTW, on OS X `make` is still version 3.81 of GNU Make.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -12,10 +12,62 @@
+ifndef CAMLYACC
+CAMLYACC := ../boot/ocamlyacc
+endif
+DESTDIR ?=
```
What does this line do?

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -12,10 +12,62 @@
+# a list of installed tools, and a function to quote a filename for
+# the shell.
+override target=''""'$(subst ','\'',$@)'#'#')'#
+override source=''""'$(subst ','\'',$<)'#'#')'#
```
This is really overkill. We don't have source (or target) files with special characters in them, and `ocamlc` would warn about them anyway. Moreover, you're also checking for the absence of spaces in file names.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -12,10 +12,62 @@
+$(and $(filter-out 1,$(words $1)),$(error \
+   cannot build file with whitespace in name))
+$1: $3 $2
+	$$(CAMLC) -compat-32 $$(LINKFLAGS) -o $$(target) $2
```
`-compat-32` is only needed for `ocamldep`. You shouldn't add it to every target without at least discussing the change.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -26,23 +78,20 @@ export OCAML_FLEXLINK:=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/flexdll/flexlink.exe
 CAMLC=$(CAMLRUN) ../boot/ocamlc -nostdlib -I ../boot -use-prims ../byterun/primitives
-CAMLOPT=$(CAMLRUN) ../ocamlopt -nostdlib -I ../stdlib
+CAMLOPT := $(CAMLRUN) ../ocamlopt
+CAMLOPT += -nostdlib -I ../stdlib
```
Why this change?

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -26,23 +78,20 @@ export OCAML_FLEXLINK:=$(ROOTDIR)/boot/ocamlrun $(ROOTDIR)/flexdll/flexlink.exe
-
-all: ocamldep ocamlprof ocamlcp ocamloptp ocamlmktop ocamlmklib dumpobj \
-     objinfo read_cmt stripdebug cmpbyt
+VPATH := $(filter-out -I,$(INCLUDES))
```
I would much rather have the dependencies with their explicit path, as ocamldep does it.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -86,46 +126,27 @@ CSLPROF_IMPORTS=misc.cmo config.cmo identifiable.cmo numbers.cmo \
-	cp ocamlcp "$(INSTALL_BINDIR)/ocamlcp$(EXE)"
-	cp ocamloptp "$(INSTALL_BINDIR)/ocamloptp$(EXE)"
-	cp profiling.cmi profiling.cmo "$(INSTALL_LIBDIR)"
+	cp -- profiling.cmi profiling.cmo "$${INSTALL_LIBDIR}"
```
This is a third way of quoting `INSTALL_LIBDIR`.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -202,7 +216,27 @@ addlabels: addlabels.cmo
-#	cp addlabels "$(INSTALL_LIBDIR)"
+#	cp addlabels "$${INSTALL_LIBDIR}"
+
+ifeq '$(EXE)' ''
```
It's not a good idea to use `EXE` to detect Windows. You should probably look at `SYSTEM` instead.
In fact, under Cygwin we have `EXE=.exe` but symbolic links are supported.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -306,68 +321,48 @@ OBJINFO=../compilerlibs/ocamlcommon.cma \
 # Copy a bytecode executable, stripping debug info
 
-STRIPDEBUG=../compilerlibs/ocamlcommon.cma \
+stripdebug=../compilerlibs/ocamlcommon.cma \
```
Why lowercase?

DemiMarie replied:
`stripdebug` is purely for internal use, not the user's.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -306,68 +321,48 @@ OBJINFO=../compilerlibs/ocamlcommon.cma \
 
 clean::
-	rm -f primreq
+	rm -f "objinfo_helper$${EXE}"
```
Why not `$(EXE)` ?

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -306,68 +321,48 @@ OBJINFO=../compilerlibs/ocamlcommon.cma \
-.ml.cmo:
-	$(CAMLC) -c $(COMPFLAGS) $<
+%.cmo: %.ml
+	$(CAMLC) -c $(COMPFLAGS) - $(source)
```
This is overkill: our source files don't have special characters.

damiendoligez commented on tools/Makefile.shared:
```diff
@@ -378,3 +373,7 @@ depend: beforedepend
 
 include .depend
+
+# Local Variables:
```
Please don't add emacs local variables to files, we use lots of different editors and lots of different emacs versions with different configurations, and for us local variables are more trouble than they're worth.

damiendoligez commented:
The makefile wizardry is a bit scary, but I like the deduplication in `tools/Makefile`.
Still, I think this should be simplified a little bit. I've added specific comments.

DemiMarie commented:
Should I squash the commits?  Many of the intermediate states don't even compile or install correctly.

shindere commented:
If the intermediate states do not even compile I think you should
definitely squash the commits so that they make sense and can be
bisected.

DemiMarie committed:
Fix AppVeyor build and Unix install when there is whitespace in paths.

damiendoligez commented on configure:
```diff
@@ -18,7 +18,17 @@
+   }
+}'
+
+configure_options="$(awk -- "$awk_script" "$@" </dev/null)" || exit 1
```
Looks like you're trying to quote the arguments, but:
1. Your script doesn't work (try an argument with a quote on a system with gawk 4.1.1).
2. We don't want to depend on `awk`.
3. It's pointless anyway because `configure_options` is only used for debugging output (modulo your addition of the `make configure` target, a lot of complexity for little benefit IMO).

FWIW, here's my own implementation of a quoting function, without `awk` (using `sed`, which we already depend on):

```
quote1(){ printf "'%s'" "`printf %s "$1" | sed -e "s/'/'\\\\\\\\''/g"`"; }

quote(){
  while : ; do
    case $# in 0) break;; esac
    quote1 "$1"
    shift
    case $# in 0) ;; *) printf " ";; esac
  done
}
```

usage: `quote1 "$1"` or `quote "$@"`

damiendoligez commented on Makefile.shared:
```diff
@@ -17,8 +17,7 @@
 # The main Makefile, fragments shared between Makefile and Makefile.nt
-
-include config/Makefile
+include global.mk
```
Please be consistent and call that `Makefile.something`.

damiendoligez commented on configure:
```diff
@@ -179,59 +235,70 @@ done
+# $(error Include $$(TOPDIR)/global.mk instead of $$(TOPDIR)/config/Makefile)\
+# endif\
+override EMPTY :=|
+' >&4
```
Several things wrong with these lines:
1. using `sed` to append to a file in a roundabout way.
2. using `\` at the end of a comment line.
3. using `EMPTY` to tell whether `global.mk` was included.

damiendoligez commented on configure:
```diff
@@ -179,59 +235,70 @@ done
+
+makefile=4
+
+exec 4>Makefile
```
Why is this better than the `>> Makefile` that was there?

damiendoligez commented on configure:
```diff
@@ -675,8 +742,8 @@ if test $with_sharedlibs = "yes"; then
-      byteccrpath="-Wl,-rpath,"
-      mksharedlibrpath="-Wl,-rpath,"
+      byteccrpath="-Wl,-rpath -Xlinker "
+      mksharedlibrpath="-Wl,-rpath -Xlinker "
```
What does this change have to do with installing the native compilers by default? Was it tested on Linux, FreeBSD, OpenBSD, NetBSD, and Haiku?

damiendoligez commented on testsuite/tests/asmcomp/mainarith.c:
```diff
@@ -91,7 +91,7 @@ void do_test(void)
       INTTEST(R[16], (X - -1));
 
-      INTTEST(R[17], ((intnat) ((char *)R - 8)));
+      INTTEST(R[17], ((intnat) ((intnat)R - 8)));
```
What?

damiendoligez commented:
There are too many changes in this PR that are not related to installing the native tools. You need to concentrate on the task at hand and leave other changes for other PRs.

DemiMarie committed:
Another attempt at fixing quoting on the Windows port

DemiMarie commented:
Can anyone help me figure out quoting on the Windows port?

DemiMarie commented:
@damiendoligez would it be okay if I closed this PR and opened a new one with only the relevant changes included?

DemiMarie committed:
Fix byterun/Makefile.nt quoting bug

DemiMarie committed:
Another attempt at a quoting fix

DemiMarie commented:
Superseded by #587.

shindere commented:
I realise this answer comes too late for this PR but for the next ones
I think it may have been better to just cleanup the branch, I don't
think it really makes sense to close the PR itself.
